### DISCUSSION TITLE
How to suppress IllegalImport?

### DISCUSSION THREAD

**Discussion by delanym** [checkstyle/checkstyle#12929] [Q&A]
With checkstyle 10.9.2 I try to suppress the `IllegalImport` check some of the time, but unable to use this approach
```xml
  <module name="SuppressWarningsFilter"/>
  <module name="TreeWalker">
    <module name="SuppressWarningsHolder"/>
    <module name="IllegalImport">
      <property name="illegalClasses" value="java.util.Hashtable"/>
    </module>
```

I placed the suppression on the **class** in a number of ways:
```java
@SuppressWarnings("illegalimport")
@SuppressWarnings("checkstyle:illegalimport")
@SuppressWarnings("checkstyle:IllegalImport")
@SuppressWarnings("check:illegalimport")
@SuppressWarnings("com.puppycrawl.tools.checkstyle.checks.import.IllegalImport")
@SuppressWarnings("com.puppycrawl.tools.checkstyle.checks.import.IllegalImportCheck")
```
but I always get `(imports) IllegalImport: Illegal import - java.util.Hashtable.`

**Comment by romani** [comment]
Try to put annotation on import.
Consider another filters to suppress violations.

https://checkstyle.org/config_filters.html

**Reply by delanym** [reply]
Annotations cant be applied to imports (tried on jdk11 and jdk20).
I also tried on the package-info.java package declaration.

I'd really rather use annotations for this. The mechanism works for other checks like https://checkstyle.org/config_modifier.html#ModifierOrder

**Reply by romani** [reply]
you need to use https://checkstyle.org/config_filters.html#SuppressionXpathSingleFilter

example: https://github.com/checkstyle/checkstyle/blob/master/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionIllegalImportTest.java#L52

```
$ cat Test.java 
import java.util.ArrayList;
import java.util.Hashtable;

public class Foo {
}

$ java -jar checkstyle-10.9.2-all.jar -t Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
|--IMPORT -> import [1:0]
|   |--DOT -> . [1:16]
|   |   |--DOT -> . [1:11]
|   |   |   |--IDENT -> java [1:7]
|   |   |   `--IDENT -> util [1:12]
|   |   `--IDENT -> ArrayList [1:17]
|   `--SEMI -> ; [1:26]
|--IMPORT -> import [2:0]
|   |--DOT -> . [2:16]
|   |   |--DOT -> . [2:11]
|   |   |   |--IDENT -> java [2:7]
|   |   |   `--IDENT -> util [2:12]
|   |   `--IDENT -> Hashtable [2:17]
|   `--SEMI -> ; [2:26]
`--CLASS_DEF -> CLASS_DEF [4:0]
    |--MODIFIERS -> MODIFIERS [4:0]
    |   `--LITERAL_PUBLIC -> public [4:0]
    |--LITERAL_CLASS -> class [4:7]
    |--IDENT -> Foo [4:13]
    `--OBJBLOCK -> OBJBLOCK [4:17]
        |--LCURLY -> { [4:17]
        `--RCURLY -> } [5:0]
```

**Reply by romani** [reply]
you can use https://checkstyle.org/config_filters.html#SuppressWithNearbyCommentFilter
configure it to be only for specific Check and message, place comment over class, and make `<property name="influenceFormat" value="-1000o"/>` to cover all possible huge imports and docs.

**Reply by delanym** [reply]
I tried SuppressionXpathSingleFilter and it works, but an annotation would be better, so I'm going with modernizer-maven-plugin that provides that. Thank you

**Reply by nrmancuso** [reply]
@delanym please share how you solved this issue with modernizer-maven-plugin

**Reply by delanym** [reply]
@nrmancuso modernizer considers java.util.HashTable a legacy api.
You can suppress with annotations (https://github.com/gaul/modernizer-maven-plugin#ignoring-elements) but I didn't do that.
I added this to my project

```xml
      <plugin>
        <groupId>org.gaul</groupId>
        <artifactId>modernizer-maven-plugin</artifactId>
        <executions>
          <execution>
            <id>modernizer</id>
            <configuration>
              <ignoreClassNamePatterns>
                <ignoreClassNamePattern>blah.blah.blah.blah.blah.blah.Classlah2</ignoreClassNamePattern>
                <ignoreClassNamePattern>blah.blah.blah.blah.blah.blah.Classlah3</ignoreClassNamePattern>
              </ignoreClassNamePatterns>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

This way the suppression is in the pom of module 43 of 500, rather than in checkstyle.xml which sits in a global build-tools artefact from another repo.

If the checkstyle.xml could be made more dynamic I might use it, but as it stands there's already so much in that file and no nice way to organize it.

**Reply by romani** [reply]
@delanym, please share exact config that worked for you

**Reply by delanym** [reply]
The parent pom has this
```xml
      <pluginManagement>
        <plugin>
          <groupId>org.gaul</groupId>
          <artifactId>modernizer-maven-plugin</artifactId>
          <version>2.6.0</version>
          <executions>
            <execution>
              <id>modernizer</id>
              <goals>
                <goal>modernizer</goal>
              </goals>
              <phase>verify</phase>
              <configuration>
                <javaVersion>19</javaVersion>
                <failOnViolations>true</failOnViolations>
              </configuration>
            </execution>
          </executions>
        </plugin>
      </pluginManagement>
```

**Reply by romani** [reply]
Sorry, I meant config of filter in checkstyle config

**Reply by delanym** [reply]
I don't have the config anymore. I went the "by non-query" route.
```xml
<module name="SuppressionXpathSingleFilter">
  <property name="files" value="FileOne.java"/>
  <property name="checks" value="IllegalImport"/>
</module>
```

**Comment by nrmancuso** [comment]
>some of the time

@delanym if there is some pattern/reason for why you need to suppress this sometimes, maybe we can help to create a filter/suppression

**Reply by delanym** [reply]
JNDI requires use of HashTable so we have to allow it https://docs.oracle.com/javase/jndi/tutorial/basics/prepare/initial.html but I don't want it used anywhere else
